File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.cpp
class Solution {
public:
    int maxPalindromes(string s, int k) {
        int n= s.length();
        int ans = 0;
        int last_end = 0;

        for(int i = 0 ; i < n ; i++){
            vector<pair<int,int>>centers = {{i,i} , {i,i+1}};
            for(auto&[l,r] : centers){
                while(l>= last_end && r < n && s[l]== s[r]){
                    if(r - l + 1 >= k){
                        ans++;
                        last_end = r +1;
                        break;
                    }
                    l--;
                    r++;
                }

                if(l >= 0 && r < n && last_end == r + 1){
                    break;
                }
            }
        }

        return ans;

    }
};

File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/2472-maximum-number-of-non-overlapping-palindrome-substrings.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        n = len(s)
        ans = 0
        last_end = 0

        for i in range(n):
            for l , r in [(i,i) , (i,i+1)]:
                while l >= last_end and r < n and s[l]==s[r]:
                    if r - l + 1 >= k:
                        ans += 1
                        last_end = r + 1
                        break
                    
                    l -= 1
                    r += 1
        
                if l >= 0 and r < n and last_end == r + 1:
                    break
        
        return ans